File: src/clientApi/java/ClientApi.java
package clientApi.java;

import java.util.List;

public class ClientApi {
    public static class Client {
        private final PersonalInfo personalInfo;
        private final List<Order> orders;
        private final boolean isInterestedInSpam;

        public Client(PersonalInfo personalInfo, List<Order> orders, boolean isInterestedInSpam) {
            this.isInterestedInSpam = isInterestedInSpam;
            this.personalInfo = personalInfo;
            this.orders = orders;
        }

        public PersonalInfo getPersonalInfo() {
            return personalInfo;
        }

        public List<Order> getOrders() {
            return orders;
        }

        public boolean isInterestedInSpam() {
            return isInterestedInSpam;
        }
    }

    public static class PersonalInfo {
        private final String name;
        private final String email;

        public PersonalInfo(String name, String email) {
            this.name = name;
            this.email = email;
        }

        public String getName() {
            return name;
        }

        public String getEmail() {
            return email;
        }
    }

    public interface Order {}
}

File: src/clientApi/kotlin/clientApi.kt
package clientApi

/*
Let's imagine we have an online shop with registered clients that can order some products.
We store the information about clients, products and order history for each client.
Also we send newsletters about new products to clients who agreed to receive it.
*/

data class ClientsData(
        val clients: List<Client>,
        val products: Set<Product>
)

data class Client(
        val personalInfo: PersonalInfo?,
        val isInterestedInSpam: Boolean,
        val orders: List<Order>
)

data class PersonalInfo(
        val name: String,
        val gender: Gender,
        val email: String?
)

data class Order(
        val product: Product,
        val status: Status
)

data class Product(
        val description: String,
        val price: Double
)

enum class Gender {
    MALE FEMALE UNSPECIFIED
}

enum class Status {
    SUCCESSFUL REJECTED IN_PROCESS
}
